protocol.go

```diff
@@ -118,13 +133,29 @@ func (p *Conn) SetWriteDeadline(t time.Time) error {
 	return p.conn.SetWriteDeadline(t)
 }
 
-func (p *Conn) readHeader() (err error) {
-	p.header, err = Read(p.bufReader)
+func (p *Conn) readHeader() error {
```

Comment:
Note that this function is called in a `once` several places -- Read, LocalAddr, RemoteAddr. If LocalAddr or RemoteAddr are called before Read, the error is ignored, which means you won't actually close the connection.

Those functions don't return an error, but probably calling them with an error value should set a flag in the conn object that tells read (or write) to exit with an error if/when they are called.

Reply:
You are right, the error shouldn’t get lost when a user asks for the addresses first before a read. Will update the PR.

Reply:
@mikey179 let us know if you can't find the time to do this. Maybe @jefferai could open a new PR that includes your commits from this branch and wrap it up? If none of you can do it, maybe I can. Thanks to both of you for your effort in collaborating 👏

Reply:
Sorry, should have mentioned that I‘m in between two jobs this month (starting a new one in February) and using the time to travel a lot, not having access to a development machine. I hope to pick this up in February again.

Reply:
Awesome! Good luck in the new adventure.

Reply:
With commit a467388 the errors aren't ignored any longer, calling `LocalAddr()` or `RemoteAddr()` before `Read()` will store the error on the connection and return it on any subsequent read.